src/lib/server/postgres_protocol_handler.cpp

```diff
@@ -0,0 +1,305 @@
+#include "postgres_protocol_handler.hpp"
+
+#include "postgres_message_types.hpp"
+
+namespace opossum {
+
+// Each message contains a field (4 bytes) indicating the packet's size including itself. Using extra variable here to
+// avoid magic numbers.
+static constexpr auto LENGTH_FIELD_SIZE = 4u;
+
+template <typename SocketType>
+PostgresProtocolHandler<SocketType>::PostgresProtocolHandler(const std::shared_ptr<SocketType>& socket)
+    : _read_buffer(socket), _write_buffer(socket) {}
+
+template <typename SocketType>
+uint32_t PostgresProtocolHandler<SocketType>::read_startup_packet() {
+  // Special SSL version number that we catch to deny SSL support
+  constexpr auto ssl_request_code = 80877103u;
+
+  const auto body_length = _read_buffer.template get_value<uint32_t>();
+  const auto protocol_version = _read_buffer.template get_value<uint32_t>();
+
+  // We currently do not support SSL
+  if (protocol_version == ssl_request_code) {
+    _ssl_deny();
+    return read_startup_packet();
+  } else {
+    // Substract uint32_t twice, since both packet length and protocol version have been read already
+    return body_length - 2 * LENGTH_FIELD_SIZE;
+  }
+}
+
+template <typename SocketType>
+PostgresMessageType PostgresProtocolHandler<SocketType>::read_packet_type() {
+  return _read_buffer.get_message_type();
+}
+
+template <typename SocketType>
+std::string PostgresProtocolHandler<SocketType>::read_query_packet() {
+  const auto query_length = _read_buffer.template get_value<uint32_t>() - LENGTH_FIELD_SIZE;
+  return _read_buffer.get_string(query_length);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::_ssl_deny() {
+  // The SSL deny packet has a special format. It does not have a field indicating the packet size.
+  _write_buffer.template put_value(PostgresMessageType::SslNo);
+  _write_buffer.flush();
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::read_startup_packet_body(const uint32_t size) {
+  // As of now, we don't do anything with the startup packet body. It contains authentication data and the
+  // database name the user desires to connect to. Hence, we read this information from the network device as
+  // one large string and throw it away.
```

Comment:
Well that sounds more like a hack. However, I think I didn't get the idea behind this.

Reply:
Imagine a benchmark with two clients that each hold a connection and separately fire transactions. Now, if we collected performance-relevant metrics within Hyrise w.r.t. the transactions from client #1 vs client #2, how would we be able to distinguish  those? We could pass above information into the metrics utility.

Reply:
I think we should completely leave this out of this PR. It might be a workaround for a problem that we do not have yet, so even adding a TODO seems too much

Reply:
My first idea to distinguish two clients is the port. However, I agree with @mrks